lucetc/src/function.rs

```diff
@@ -78,6 +84,138 @@ impl<'a> FuncInfo<'a> {
+                i64::from(*environ.op_offsets.last().unwrap()),
+            );
+            let new_instr_count = builder.ins().iadd(cur_instr_count, update_const.into());
+            builder
+                .ins()
+                .store(flags, new_instr_count, addr, instr_count_offset);
+            *environ.op_offsets.last_mut().unwrap() = 0;
+        };
+
+        // Update the instruction counter, if necessary
+        let op_cost = match op {
+            // Opening a scope is a syntactic operation, and free.
+            Operator::Block { .. } |
+            // These do not add counts, see above comment about return/unreachable
+            Operator::Unreachable |
+            Operator::Return => 0,
+            // Call is quick
+            Operator::Call { .. } => 1,
+            // but indirect calls take some extra work to validate at runtime
+            Operator::CallIndirect { .. } => 2,
+            // Testing for an if involve some overhead, for now say it's also 1
+            Operator::If { .. } => 1,
+            // Else is a fallthrough or alternate case for something that's been tested as `if`, so
+            // it's already counted
+            Operator::Else => 0,
+            // Entering a loop is a syntactic operation, and free.
+            Operator::Loop { .. } => 0,
+            // Closing a scope is a syntactic operation, and free.
+            Operator::End => 0,
+            // Taking a branch is an operation
+            Operator::Br { .. } => 1,
+            // brif might be two operations?
+            Operator::BrIf { .. } => 1,
+            // brtable is kind of cpu intensive compared to other wasm ops
+            Operator::BrTable { .. } => 2,
+            // nop and drop are free
+            Operator::Nop |
+            Operator::Drop => 0,
+            // everything else, just call it one operation.
+            _ => 1,
+        };
+        self.op_offsets.last_mut().map(|x| *x += op_cost);
+
+        // apply flushing behavior if applicable
+        match op {
+            Operator::Unreachable
+            | Operator::Return
+            | Operator::CallIndirect { .. }
+            | Operator::Call { .. }
+            | Operator::Block { .. }
+            | Operator::Loop { .. }
+            | Operator::If { .. }
+            | Operator::Else
+            | Operator::Br { .. }
+            | Operator::BrIf { .. }
+            | Operator::BrTable { .. } => {
+                flush_counter(self, builder);
+            }
+            Operator::End => {
+                // We have to be really careful here to avoid violating a cranelift invariant:
+                // if the next operation is `End` as well, this end will have marked the block
+                // finished, and attempting to add instruction to update the instruction counter
+                // will cause a panic.
+                //
+                // We can avoid that case by ensuring instruction counts are flushed at the *entry*
```

Comment:
Is that what we're doing here, or is that something we need to do in the future? It looks like we're just calling `flush_counter` in the same way we do for the other basic block terminators

Reply:
oops, good eye. this is a relic from my initial attempt at this, and currently will call `flush_counter`, but the counter is always 0 (so no instructions are added, and no cranelift invariants are violated)

I'll reword this and change the call to an assert

Reply:
(several hours of thought later) I was wrong - the counter is _definitely_ not always 0 at this point, and I reworded this comment and added an example of what it's talking about for reference.